common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java

```diff
@@ -300,7 +300,7 @@ public ShuffleMetrics() {
     }
 
     ManagedBufferIterator(FetchShuffleBlocks msg, int numBlockIds) {
-      final int[] mapIdAndReduceIds = new int[2 * numBlockIds];
+      final long[] mapIdAndReduceIds = new long[2 * numBlockIds];
```

Comment:
maybe we should one long[] for map id and one int[] for reduce id.

Reply:
Actually we already have long[] for map id and int[] for reduce id in the message, here we need is kinda assemble work to flatten reduce id and its corresponding mapid.
The current way waste memory, we can also do it in a cpu consuming way, which is for each index, calculate which map id and reduce id corresponding with the `idx`.

Reply:
After taking a further look, I split the new protocol managed buffer iterator in 
539d725, that make us more flexible to control the iterator and no more array created.